### PR TITLE
Rust: Implement `UnusedValue.ql`

### PR DESCRIPTION
**Pull request by hvitved**
This PR implements an initial version of `UnusedValue.ql`, based on SSA.

### CODE REVIEW

File: rust/ql/src/queries/unusedentities/UnusedValue.ql [hunk 1]
```diff
@@ -9,7 +9,16 @@
...
+import codeql.rust.dataflow.internal.SsaImpl
+import UnusedVariable
 
-from Locatable e
-where none() // TODO: implement query
-select e, "Variable is assigned a value that is never used."
+from AstNode write, Ssa::Variable v
+where
+  variableWrite(write, v) and
+  // SSA definitions are only created for live writes
```
Comment (geoffw0) [inline rust/ql/src/queries/unusedentities/UnusedValue.ql:19]: What is a "live write"?

File: rust/ql/test/query-tests/unusedentities/main.rs [hunk 2]
```diff
@@ -189,7 +189,7 @@ enum YesOrNo {
     No,
 }
 
-use YesOrNo::{Yes, No}; // allows `Yes`, `No` to be accessed without qualifiers.
+use YesOrNo::{No, Yes}; // allows `Yes`, `No` to be accessed without qualifiers.
```
Comment (geoffw0) [inline rust/ql/test/query-tests/unusedentities/main.rs:192]: Why this change?

File: rust/ql/test/query-tests/unusedentities/main.rs [hunk 3]
```diff
@@ -222,25 +223,22 @@ fn if_lets_matches() {
...
+        None => {}
     }
 
     let d = Some(70);
     match d {
         Some(val) => {
-            total += val;
-        }
-        None => {
+            total += val; // BAD: unused variable
```
Comment (geoffw0) [inline rust/ql/test/query-tests/unusedentities/main.rs:232]:
```suggestion
            total += val; // BAD: unused value
```

File: rust/ql/src/queries/unusedentities/UnusedValue.ql (hunk 1 above)
Comment (hvitved) [reply rust/ql/src/queries/unusedentities/UnusedValue.ql:19]: One that can reach a read.

File: rust/ql/test/query-tests/unusedentities/main.rs (hunk 2 above)
Comment (hvitved) [reply rust/ql/test/query-tests/unusedentities/main.rs:192]: Hmm, I think it's the auto-formatter.

Comment (geoffw0) [reply rust/ql/test/query-tests/unusedentities/main.rs:192]: In general I believe we should not autoformat test code in the target language.  Autoformatting adds a regularity that could (theoretically) mask issues in the extractor.

Comment (geoffw0) [reply rust/ql/test/query-tests/unusedentities/main.rs:192]: Not that I strongly care about this particular case (reverting not necessary).

File: rust/ql/src/queries/unusedentities/UnusedValue.ql (hunk 1 above)
Comment (geoffw0) [reply rust/ql/src/queries/unusedentities/UnusedValue.ql:19]: Ah of course, the key line of the query.

File: rust/ql/test/query-tests/unusedentities/main.rs (hunk 3 above)
Comment (hvitved) [reply rust/ql/test/query-tests/unusedentities/main.rs]: Thanks, fixed.

**Review by geoffw0 — COMMENTED** [review]
LGTM once https://github.com/github/codeql/pull/17713 is merged.

**Comment by geoffw0** [comment]
The underlying PR (first two commits here) is merged and I'm happy with the third commit.  However there are merge conflicts and we need a DCA run, but @hvitved is away.  So I've made a new PR here containing the third commit + merge fixup here: https://github.com/github/codeql/pull/17773

**Comment by hvitved** [comment]
Superseded by https://github.com/github/codeql/pull/17773.
